modules/cdp-engine/src/lib.rs

```diff
@@ -1216,59 +1272,212 @@ impl<T: Config> Pallet<T> {
 			.reciprocal()
 			.unwrap_or_else(Ratio::max_value)
 			.saturating_mul_int(
 				T::PriceSource::get_relative_price(T::GetStableCurrencyId::get(), currency_id)
-					.expect("the oracle price should be avalible because liquidation are triggered by it.")
+					.expect("the oracle price should be available because liquidation are triggered by it.")
 					.saturating_mul_int(target_stable_amount),
 			);
 		let collateral_supply = amount.min(max_supply_limit);
 
-		// try swap collateral to stable to settle debit swap succeed.
-		if let Ok((actual_supply_collateral, actual_target_amount)) =
-			<T as Config>::CDPTreasury::swap_collateral_to_stable(
-				currency_id,
-				SwapLimit::ExactTarget(collateral_supply, target_stable_amount),
-				false,
-			) {
-			let refund_collateral_amount = amount
-				.checked_sub(actual_supply_collateral)
-				.expect("swap success means collateral >= actual_supply_collateral; qed");
-
-			// refund remain collateral to CDP owner
-			if !refund_collateral_amount.is_zero() {
-				<T as Config>::CDPTreasury::withdraw_collateral(who, currency_id, refund_collateral_amount)?;
-			}
+		let (actual_supply_collateral, actual_target_amount) = <T as Config>::CDPTreasury::swap_collateral_to_stable(
+			currency_id,
+			SwapLimit::ExactTarget(collateral_supply, target_stable_amount),
+			false,
+		)?;
 
-			// Note: for StableAsset, the swap of cdp treasury is always on `ExactSupply`
-			// regardless of this swap_limit params. There will be excess stablecoins that
-			// need to be returned to the `who` from cdp treasury account.
-			if actual_target_amount > target_stable_amount {
-				<T as Config>::CDPTreasury::withdraw_surplus(
-					who,
-					actual_target_amount.saturating_sub(target_stable_amount),
-				)?;
-			}
-		} else {
-			// if cannot liquidate by swap, create collateral auctions by cdp treasury
-			<T as Config>::CDPTreasury::create_collateral_auctions(
-				currency_id,
-				amount,
-				target_stable_amount,
-				who.clone(),
-				true,
+		let refund_collateral_amount = amount
+			.checked_sub(actual_supply_collateral)
+			.expect("swap success means collateral >= actual_supply_collateral; qed");
+		// refund remain collateral to CDP owner
+		if !refund_collateral_amount.is_zero() {
+			<T as Config>::CDPTreasury::withdraw_collateral(who, currency_id, refund_collateral_amount)?;
+		}
+
+		// Note: for StableAsset, the swap of cdp treasury is always on `ExactSupply`
+		// regardless of this swap_limit params. There will be excess stablecoins that
+		// need to be returned to the `who` from cdp treasury account.
+		if actual_target_amount > target_stable_amount {
+			<T as Config>::CDPTreasury::withdraw_surplus(
+				who,
+				actual_target_amount.saturating_sub(target_stable_amount),
 			)?;
 		}
 
 		Ok(())
 	}
-	pub fn get_collateral_currency_ids() -> Vec<CurrencyId> {
-		CollateralParams::<T>::iter_keys().collect()
+}
+
+pub struct LiquidateViaContracts<T>(PhantomData<T>);
```

Comment:
@shaunxw I think this needs to have a gasLimit or blockWeight ratio and we can iterate until OOG

Reply:
Currently for each contract call there is a gas limit (and storage limit) set, like for `liquidate` https://github.com/AcalaNetwork/Acala/blob/ce08df858d3bdb08e7c6b3c5f201ab3bc5dc77a1/modules/evm-bridge/src/lib.rs#L282-L289

Additionally there is a max limit for the number of contracts. I think these should do the work?

Reply:
yeah if we manage the limit then yes, but I still think we need to make this smarter. how do we know the gas limit is enough? do we need to retry with more gas? I think we can do some improvements

Reply:
We don't need to retry. Smart contracts need to be implemented to meet gas limit requirements.